perf/perf.go

```diff
@@ -0,0 +1,529 @@
+	case unix.ENOSYS:
+		return -1, errors.WithMessage(unix.ENOSYS, "sample type not supported by the hardware")
+	case unix.EOPNOTSUPP:
+		return -1, errors.WithMessage(unix.EOPNOTSUPP, "this event is not supported by the hardware or requires a feature not supported by the hardware")
+	case unix.EOVERFLOW:
+		return -1, errors.WithMessage(unix.EOVERFLOW, "sample_max_stack is larger than the kernel support; check \"/proc/sys/kernel/perf_event_max_stack\" for maximum supported size")
+	case unix.EPERM:
+		return -1, errors.WithMessage(unix.EPERM, "insufficient capability to request exclusive access")
+	case unix.ESRCH:
+		return -1, errors.WithMessage(unix.ESRCH, "pid does not exist")
+	default:
+		return -1, err
+	}
+}
+
+func addToEpoll(epollfd, fd int, cpu int) error {
+	if int64(cpu) > math.MaxInt32 {
+		return errors.Errorf("unsupported CPU number: %d", cpu)
+	}
+
+	// The representation of EpollEvent isn't entirely accurate.
+	// Pad is fully useable, not just padding. Hence we stuff the
+	// CPU in there, which allows us to use a slice to access
+	// the correct perf ring.
+	event := unix.EpollEvent{
+		Events: unix.EPOLLIN,
+		Fd:     int32(fd),
+		Pad:    int32(cpu),
+	}
+
+	err := unix.EpollCtl(epollfd, unix.EPOLL_CTL_ADD, fd, &event)
+	return errors.Wrap(err, "can't add fd to epoll")
+}
+
+func cpuForEvent(event *unix.EpollEvent) int {
+	return int(event.Pad)
+}
+
+func (ring *perfEventRing) Close() {
+	runtime.SetFinalizer(ring, nil)
+	unix.Close(ring.fd)
+	unix.Munmap(ring.mmap)
+
+	ring.fd = -1
+	ring.mmap = nil
+}
+
+func (ring *perfEventRing) LoadHead() {
+	ring.reader.loadHead()
+}
+
+// NB: Has to be preceded by a call to LoadHead.
+func (ring *perfEventRing) ReadRecord() (*Record, error) {
+	if ring.fd == -1 {
+		return nil, errClosed
+	}
+
+	defer ring.reader.writeTail()
+	return readRecord(ring.reader)
+}
+
+// Record contains either a sample or a counter of the
+// number of lost samples.
+type Record struct {
+	// The data submitted via bpf_perf_event_output.
+	// They are padded with 0 to have a 64-bit alignment.
+	// If you are using variable length samples you need to take
+	// this into account.
+	RawSample []byte
+
+	// The number of samples which could not be output, since
+	// the ring buffer was full.
+	LostSamples uint64
+}
+
+func readRecord(rd io.Reader) (*Record, error) {
+	const (
+		perfRecordLost   = 2
+		perfRecordSample = 9
+	)
+
+	var header perfEventHeader
+	err := binary.Read(rd, internal.NativeEndian, &header)
+	if err == io.EOF {
+		return nil, nil
+	}
+
+	if err != nil {
+		return nil, errors.Wrap(err, "can't read event header")
+	}
+
+	switch header.Type {
+	case perfRecordLost:
+		lost, err := readLostRecords(rd)
+		return &Record{LostSamples: lost}, err
+
+	case perfRecordSample:
+		sample, err := readRawSample(rd)
+		return &Record{RawSample: sample}, err
+
+	default:
+		return nil, errors.Errorf("unknown event type %d", header.Type)
```

Comment:
I wondered how we want user libraries to handle the variety of events that we might generate here as well. It might be kinda neat to have users do something like:

```
event, err := reader.Read()
if err {
  ...
  continue
}
switch event.(type) {
case SampleRecord:
...
case LostRecord:
...
case UnknownRecord:
...
}
```

Reply:
What do you folks do to handle the different types of events that might be generated here?

Reply:
As far as I can tell, because of
```
		Type:        unix.PERF_TYPE_SOFTWARE,
		Config:      unix.PERF_COUNT_SW_BPF_OUTPUT,
```
We can only receive raw events from the ring. I think we've never seen the unknown record being triggered. Do you have a different set up?

Reply:
Re: your idea with `SampleRecord`, etc. It forces us to return an interface, which will make it impossible to avoid allocations.

So far my hope is that we can just add more members to `Record` if we need to. Not great, admittedly. The other option is to introduce completely distinct Record types, and add `Read` methods for them. Also not great.

Reply:
I was wondering how allocations might fall into this, good point. There's not many record types anyway so it's not a big deal for that interface idea, we can drop that.

Reply:
> We can only receive raw events from the ring. I think we've never seen the unknown record being triggered. Do you have a different set up?

I've never seen it personally but there are rumours in the office that these were seen last year some time ;)

My main concern is to avoid logging the same error message every time we receive an event like this and causing disk issues due to this. So long as we have a way to tell that a given error is the same as the one we saw before, then we will have a reasonable means to ratelimit logging. Whether that's by having a static error variable that we can check against or a function that we can use to check "is it X type of error?" or we present unknown errors like this as a `Record{Unknown: true}` (or some variation) I don't have a strong opinion.

Currently as returned right now it looks like a brand new error pointer each time and there's no sense of "identity" for this error, so no way to tell from a library user whether this is the kind of error it is. Maybe we can just add a new error class for this error type.

Reply:
I've added `IsUknownEvent(error)`.